src/test/java/com/hivemq/mqtt/message/pool/FreeIdRangesTest.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2019-present HiveMQ GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hivemq.mqtt.message.pool;
+
+import com.google.common.collect.Lists;
+import com.hivemq.extension.sdk.api.annotations.NotNull;
+import com.hivemq.mqtt.message.pool.exception.NoMessageIdAvailableException;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.hivemq.bootstrap.ClientConnection.MAX_MESSAGE_ID;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public class FreeIdRangesTest {
+
+    @Test
+    public void takeNextId_whenTakingIdsSequentiallyAndReturning_thenSequentialIdsAreProvided() throws
+            NoMessageIdAvailableException {
+
+        final List<Integer> integers = new ArrayList<>();
+        final FreeIdRanges messageIDPool = new FreeIdRanges(1, MAX_MESSAGE_ID);
+
+        for (int i = 0; i < MAX_MESSAGE_ID; i++) {
+            final int id = messageIDPool.takeNextId();
+            integers.add(id);
+        }
+
+        for (final Integer id : integers) {
+            messageIDPool.returnId(id);
+        }
+
+        for (int i = 0; i < MAX_MESSAGE_ID; i++) {
+            final int id = messageIDPool.takeNextId();
+            integers.add(id);
+        }
+
+        assertTrue(areConsecutiveMessageIds(Lists.partition(integers, MAX_MESSAGE_ID).get(0)));
+        assertTrue(areConsecutiveMessageIds(Lists.partition(integers, MAX_MESSAGE_ID).get(1)));
+    }
+
+    @Test(expected = NoMessageIdAvailableException.class)
+    public void takeNextId_whenNoMoreIdsAvailable_thenReturnSpecialResult() throws
```

Comment:
`thenReturnSpecialResult` Is an exception a "special result"? 😄

Reply:
haha, a remnant from the iteration where I used -1 instead of exceptions